Src/PuTTY.c

```diff
@@ -0,0 +1,365 @@
+
+#include "usart.h"
+#include "PuTTY.h"
+#include <stdint.h>
+#include "main.h"
+#include "gpio.h"
+
+///////////////////////////////////////////////////// DEFINITIONS
+// Commands to be remember
+#define huartx huart3           // What huart communication is used
+#define COMMANDS_TO_REMEMBER 16 // Used in PC commands 2D matrix
+#define MAX_COMMAND_LENGTH 32
+
+///////////////////////////////////////////////////// PRIVATE FUNCTION DECLARATIONS
+static void Putty_TextOut(char *str);                                // Displays the code on the console
+static void Putty_HexOut(uint8_t data[], uint8_t length);            // Transmits data to PC
+static void Putty_HandleCommand(char *input);                        // Executes action depending on input command
+static void Putty_HandlePcInput(char *input, size_t n_chars);        // Called with timer | Change this
+static uint8_t Putty_Wrap(uint8_t val, int8_t dif, uint8_t modulus); // Keeps values within the real range
+static void Putty_ClearLine();                                       //Clears the current line to that new text can be placed.
+
+
+
+///////////////////////////////////////////////////// PUBLIC FUNCTION IMPLEMENTATIONS
+Putty_Enum Putty_Init()
+{
+    Putty_Vars.errorCode = 0;
+    Putty_Vars.huart_Rx_len = 0;
+
+    char *startmessage = "----------PuttyInterface_Init-----------\n\r"; // Initial message
+    Putty_printf(startmessage);
+
+    HAL_UART_Receive_IT(&huartx, Putty_Vars.rec_buf, 1); // Data reception under serial interrupt mode
+
+    return Putty_Vars.errorCode = NoErrors;
+}
+
+Putty_Enum Putty_GetErrorCode()
+{
+    return Putty_Vars.errorCode;
+}
+
+Putty_Enum Putty_printf(char *input)
+{
+    sprintf(Putty_Vars.smallStrBuffer, input); // Copies and turns into string
+    Putty_TextOut(Putty_Vars.smallStrBuffer);  // Outputs to console
+    return Putty_Vars.errorCode = NoErrors;
+}
+
+Putty_Enum Putty_DeInit()
+{
+    // TODO properly
+    return Putty_Vars.errorCode = NoErrors;
+}
+
+Putty_Enum Putty_Reset()
+{
+    Putty_Init(Putty_Vars);
+    return Putty_Vars.errorCode = NoErrors;
+}
+
+// Currently in main, but call it in a timer please.
+// Note. Used to be called PuttyInterface_Update
+Putty_Enum Putty_Callback()
+{
+    // If a command has been received
+    if (Putty_Vars.huart_Rx_len)
+    {
+        Putty_HandlePcInput((char *)&Putty_Vars.small_buf, Putty_Vars.huart_Rx_len); // Deal with the PC Input
+        Putty_Vars.huart_Rx_len = 0;                                                  // Reset the size of the input
+        HAL_UART_Receive_IT(&huartx, Putty_Vars.rec_buf, 1);                          // Keep on checking for new data
+    }
+    return Putty_Vars.errorCode = NoErrors;
+}
+
+Putty_Enum Putty_UARTCallback(UART_HandleTypeDef *huart)
+{
+    Putty_Vars.huart_Rx_len = 1;                        // Reset the length received
+    Putty_Vars.small_buf[0] = *(huart->pRxBuffPtr - 1); // Copies received data into buffer
```

Comment:
what happens if multiple callbacks are received before it is processed, it looks like it is overwritten